spec/relativetimeformat.html

```diff
@@ -1,61 +1,256 @@
+          1. Let _unit_ be GetBestMatchUnit(_units_).
+        1. Else
+          1. Let _unit_ be _relativeTimeFormat_.[[unit]].
+        1. If _style_ is equal to "short", then
+          1. Let _entry_ be _unit_ + "-short".
+        1. Else if _style_ is equal to "narrow", then
+          1. Let _entry_ be _unit_ + "-narrow".
+        1. Else
+          1. Let _entry_ be _unit_.
+        1. Let _exists_ be ? HasProperty(_fields_, _entry_).
+        1. If _exists_ is *false*, then
+          1. Let _entry_ be _unit_.
+        1. Let _patterns_ be ? Get(_fields_, _entry_).
+        1. Let _v_ be _units_.[[&lt;_unit_&gt;]].
+        1. If _v_ is between -2 (inclusive) and 2 (inclusive), then
+          1. Let _exists_ be ? HasProperty(_patterns_, ToString(_v_)).
+          1. If _exists_ is *true*, then
+            1. Let _result_ be Get(_patterns_, ToString(_v_)).
+            1. Return _result_.
+        1. If _v_ is less than 0, then
+          1. Let _tl_ be "past".
+        1. Else
+          1. Let _tl_ be "future".
+        1. Let _po_ be ? Get(_patterns_, _tl_).
+        1. Let _fv_ be _relativeTimeFormat_.[[NumberFormat]](_v_).
+        1. Let _pr_ be _relativeTimeFormat_.[[PluralRule]](_v_).
+        1. Let _pattern_ be ? Get(_po_, _pr_).
+        1. Let _values_ be a new Record;
+        1. Set _values_["0"] to be new Record { [[type]]: "number", [[value]]: _fv_ }.
+        1. Return DeconstructPattern(_pattern_, _values_).
+      </emu-alg>
+    </emu-clause>
+
+    <emu-clause id="sec-deconstructpattern" aoid="DeconstructPattern">
+      <h1>DeconstructPattern (pattern, placeables)</h1>
+
+      <p>
+        The *DeconstructPattern* abstract operation is called with arguments _pattern_ (which must be a String) and _placeables_ (which must be a Record), and deconstructts the pattern string into a list of parts.
+        Example:
+        <pre>
+        Input:
+          DeconstructPattern("AA{xx}BB{yy}CC", {
+            xx: {type: 'hour', value: '15'},
+            yy: {type: 'minute', value: '06'}
+          });
+
+        Output (List of Records):
+          [
+            {type: 'literal', value: 'AA'},
+            {type: 'hour', value: '15'},
+            {type: 'literal', value: 'BB'},
+            {type: 'minute', value: '06'},
+            {type: 'literal', value: 'CC'}
+          ]
+        </pre>
+      </p>
+      <emu-alg>
+        1. Set _parts_ to be the result of _pattern_.split(/\{([^\}]+)\}/).
+        1. Let _result_ be a new empty List.
+        1. Set _i_ to be *0*.
+        1. Repeat, while _i_ &lt; _parts_.length
+          1. Set _part_ to be _parts_.[_i_].
+          1. If _i_ % 2 is *0*, then
+            1. If _part_.length is &gt; *0*, then
+              1. Add new part record { [[type]]: *"literal"*, [[value]]: _part_ } as a new element on the list _result_.
+          1. Else,
+            1. Set _subst_ to _placeables_.[[_part_]].
+            1. If !subst, then
+              1. Throw new Error.
```

Comment:
yeah, we'll need to figure out the exact exception, I'd prefer to do this once we finalize the `DeconstructPattern` algo.